lib/virtmanager.pm

```diff
@@ -584,4 +584,27 @@ sub create_guest {
     assert_and_click 'virtman_guest_begin_install';
 }
 
+# Expecting to see the guest screen
+sub detect_login_screen {
+    my $timeout = shift // 5;
+    if (!check_screen 'virt-manager_login-screen', $timeout) {
+        wait_still_screen 3;    # Connecting to guest's console
+        mouse_set(30, 200);     # Go inside of the guest's console
+        mouse_click();
+        save_screenshot();
+        hold_key "ctrl-alt";       # Escape from the guest's console
+        release_key "ctrl-alt";    # Now the mouse pointer is free
+        mouse_set(300, 70);
+    }
+    if (!check_screen 'virt-manager_login-screen', 5) {
+        send_key 'esc';
+        send_key 'backspace';
+    }
+    if (!check_screen 'virt-manager_login-screen', 5) {
+        assert_and_click 'virt-manager_send-key';
+        assert_and_click 'virt-manager_ctrl-alt-f2';
+        assert_screen "virt-manager_login-screen";
+    }
```

Comment:
In both above cases I suggest to use a multi-tag `assert_screen` with `match_has_tag` instead of `check_screen` with non-zero timeout to prevent introducing any timing dependant behaviour, to save test execution time as well as state more explicitly from the testers point of view what are the expected alternatives. For example:

```perl
assert_screen([qw(yast2_console-finished yast2_missing_package)]);
if (match_has_tag('yast2_missing_package')) {
    send_key 'alt-o';  # confirm package installation
    assert_screen 'yast2_console-finished';
}
```

Reply:
Well here I deal with a VM which should present the login screen (either graphical or text one).
Sometimes the screen is not present due to black screen / screensaver / etc...

My point is: I check for only **one tag** but I do it **until it is matched**.
Then the first timeout can be modified in case the VM is f.e. rebooting.

I don't know how to implement your suggestion.

Reply:
Well, I think you already phrased it exactly as how I would write it :)

Please take a look into the example in
https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/x11/kontact.pm#L37

Reply:
I'm sorry, in that example a block of code is executed again and again until tag is matched, but in my case different block of code are executed until tag is matched plus the timeout of the first attempt can be changed.
```

Reply:
```perl
sub detect_login_screen {
    my $timeout = shift // 5;
    if (!check_screen 'virt-manager_login-screen', $timeout) {
        wait_still_screen 3;    # Connecting to guest's console
        mouse_set(30, 200);     # Go inside of the guest's console
        mouse_click();
        save_screenshot();
        hold_key "ctrl-alt";       # Escape from the guest's console
        release_key "ctrl-alt";    # Now the mouse pointer is free
        mouse_set(300, 70);
    }
    if (!check_screen 'virt-manager_login-screen', 5) {
        send_key 'esc';
        send_key 'backspace';
    }
    if (!check_screen 'virt-manager_login-screen', 5) {
        assert_and_click 'virt-manager_send-key';
        assert_and_click 'virt-manager_ctrl-alt-f2';
        assert_screen "virt-manager_login-screen";
    }
}
```